WIP: possible organization for benchmarking

davidagold opened the pull request:
This PR explores a possible means of organizing benchmarking code. In particular, it uses the `Benchmarks@benchmarkable` as a groundwork for writing the results of profiling tests to tsv documents in the repo. An example of this organization can be found in `issues/fastmathcomplex.jl`.

The main idea is that one defines an inner "test" method (here `_fastmathcomplex_test`) that contains just the code to be benchmarked. One then passes this method as the `core` parameter to `Benchmarks.@benchmarkable` to produce the main benchmarking method (here `_fastmathcomplex`). We then wrap this benchmarking method in a comprehensive profiling function (here `profile_fastmathcomplex`) that grabs the environment information, initializes a `Benchmarks.Samples` object, passes the latter to the main benchmarking method and writes the results and the environment information to a document in a `res` folder.

IainNZ commented on issues/fastmathcomplex.jl:
```diff
@@ -7,19 +7,36 @@
+    env = Benchmarks.Environment()
+    _fastmathcomplex(samples, n_samples, n_evals)
+    pkg_dir = Pkg.dir("BasePerfTests")
+    writecsv(pkg_dir * "res/env.tsv", env)
```
joinpath

IainNZ commented:
Wow, this is really good. I think we should go down this path. Would you like commit access?

davidagold commented:
Yes, I would love to help contribute to this project, thank you!

I'll incorporate your suggestion and rebase. Also, before anything gets merged we may as well think about a couple of things:

1. tsv or csv?
2. Do we want a separate file for each writeout, or do we want a single file that contains environment information, a single file that contains profiling results as well as the name of each profiler (e.g. `fastmathcomplex`)? We could also include a third file that contains information about the profilers themselves, such as the commit at which they were added to BasePerfTests and the methods from Base that are tested in the profiler. These files altogether would essentially constitute tables from which one could grab, for a given benchmarking session, the details of the environment of the test and the methods involved in the test.
3. Is there any other information that ought to be included for the benchmark results table? E.g. an `n` field to be used for estimating complexity?

IainNZ commented:
I've generally used CSV, I don't think we'll be storing much text so I don't think it matters too much.

I'll have to think about the rest. The goal of the whole thing in my mind is to feed into a web interface that gives you some idea, across time, how performance has changed. So it'd need stuff like date, Julia version, and yeah probably best to store the parameters that were used and the BasePerfTests commit used.

davidagold commented:
@IainNZ I've refactored the code to include a generic `profile` method that takes a benchmarking method, the number of evaluations per sample and the number of samples as arguments. This has allowed for a slight simplification of the method nomenclature, though I'm not wedded to the latter. I've applied the scheme to the five original issues in `/issues`. I've included a `run_benchmarks.jl` file to function similarly to `runtests.jl`. Each issue is sandboxed in its own module.

Finally, I've implemented the repetition of the method to be profiled in `fastmathcomplex.jl` and `highdimsetindex.jl` in terms of the `n_evals` rather than the explicit loop inside the test method. This is perhaps the biggest change.

My git-fu is still developing, so bear with me as I try to resolve whatever conflicts there are. In the meantime, do you have any thoughts on these developments?

IainNZ commented:
Wow looks really nice. I think this is definitely a great direction to be going in.
